samcli/lib/sync/infra_sync_executor.py

```diff
@@ -0,0 +1,295 @@
+    _deploy_context: DeployContext
+
+    def __init__(self, build_context: BuildContext, package_context: PackageContext, deploy_context: DeployContext):
+        """Constructs the sync for infra executor.
+
+        Parameters
+        ----------
+        build_context : BuildContext
+            BuildContext
+        package_context : PackageContext
+            PackageContext
+        deploy_context : DeployContext
+            DeployContext
+        """
+        self._build_context = build_context
+        self._package_context = package_context
+        self._deploy_context = deploy_context
+
+        session = Session(profile_name=self._deploy_context.profile, region_name=self._deploy_context.region)
+        self._cfn_client = self._boto_client("cloudformation", session)
+        self._s3_client = self._boto_client("s3", session)
+
+    def _boto_client(self, client_name: str, session: Session):
+        """
+        Creates boto client
+
+        Parameters
+        ----------
+        client_name: str
+            The name of the client
+        session: boto3.Session
+            The session created using customer config
+
+        Returns
+        -------
+        Service client instance
+        """
+        return get_boto_client_provider_from_session_with_config(session)(client_name)
+
+    def _compare_templates(self, local_template_path: str, stack_name: str) -> bool:
+        """
+        Recursively conpares two templates, including the nested templates referenced inside
+
+        Parameters
+        ----------
+        local_template_path : str
+            The local template location
+        stack_name : str
+            The CloudFormation stack name that the template is deployed to
+
+        Returns
+        -------
+        bool
+            Returns True if two templates are identical
+            Returns False if two templates are different
+        """
+        current_template = None
+        # If the customer template uses a nested stack with location/template URL in S3
+        if local_template_path.startswith("https://"):
+            parsed_s3_location = re.search(r"https:\/\/[^/]*\/([^/]*)\/(.*)", local_template_path)
+            if parsed_s3_location:
+                s3_bucket = parsed_s3_location.group(1)
+                s3_key = parsed_s3_location.group(2)
+                try:
+                    s3_object = self._s3_client.get_object(Bucket=s3_bucket, Key=s3_key)
+                except ClientError as ex:
+                    LOG.debug("The provided template location %s can not be found", local_template_path, exc_info=ex)
+                    return False
+
+                streaming_body = s3_object.get("Body")
+                if streaming_body:
+                    current_template = yaml_parse(streaming_body.read().decode("utf-8"))
+
+        # If the template location is local
+        else:
+            current_template = get_template_data(local_template_path)
```

Comment:
This part until here can be moved to a separate function to improve the readability.